libs/platform/ebpf_ring_buffer.h

```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * @brief Mark one or more records in the ring buffer as returned to the ring.
+     *
+     * @param[in,out] ring_buffer Ring buffer to update.
+     * @param[in] length Length of bytes to return to the ring buffer.
+     * @retval EPBF_SUCCESS Successfully returned records to the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to return records to the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_return(_Inout_ ebpf_ring_buffer_t* ring_buffer, size_t length);
+
+    /**
+     * @brief Get pointer to the ring buffer shared data.
+     *
+     * @param[in] ring_buffer Ring buffer to query.
+     * @param[out] buffer Pointer to ring buffer data.
+     * @retval EPBF_SUCCESS Successfully mapped the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to map the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_map_buffer(_In_ const ebpf_ring_buffer_t* ring_buffer, _Outptr_ uint8_t** buffer);
+
+    /**
+     * @brief Reserve a buffer in the ring buffer. Buffer is valid until either ebpf_ring_buffer_submit,
+     * ebpf_ring_buffer_discard, or the end of the current epoch.
+     *
+     * @param[in,out] ring_buffer Ring buffer to update.
+     * @param[out] data Pointer to start of reserved buffer on success.
+     * @param[in] length Length of buffer to reserve.
+     * @retval EPBF_SUCCESS Successfully reserved space in the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to reserve space in the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_reserve(_Inout_ ebpf_ring_buffer_t* ring_buffer, _Outptr_ uint8_t** data, size_t length);
+
+    /**
+     * @brief Mark a previously reserved buffer as available.
+     *
+     * @param[in] data Pointer to buffer to submit.
+     * @retval EPBF_SUCCESS Successfully submitted space in the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to submit space in the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_submit(_Frees_ptr_opt_ uint8_t* data);
+
+    /**
+     * @brief Mark a previously reserved buffer as discarded.
+     *
+     * @param[in] data Pointer to buffer to submit.
+     * @retval EPBF_SUCCESS Successfully discarded space in the ring buffer.
+     * @retval EBPF_INVALID_ARGUMENT Unable to discard space in the ring buffer.
+     */
+    ebpf_result_t
+    ebpf_ring_buffer_discard(_Frees_ptr_opt_ uint8_t* data);
+
+    /**
+     * @brief Locate the next record in the ring buffer's data buffer and
+     * advance consumer offset.
+     *
+     * @param[in] buffer Pointer to the start of the ring buffer's data buffer.
+     * @param[in] buffer_length Length of the ring buffer's data buffer.
+     * @param[in] consumer Consumer offset.
+     * @param[in] producer Producer offset.
+     * @return Pointer to the next record or NULL if no more records.
+     */
+    const ebpf_ring_buffer_record_t*
```

Comment:
This is the only API dealing with record. Why make record public at all? Why not hide it as an implementation detail and let callers deal with raw buffers only?

Reply:
The entire ring buffer is exposed to user mode. Applications can write their own logic to consume entries from it, for which they need to know the format of the header.